### PR TITLE
Add support for switch expressions in dataflow

### LINKED ISSUE #2373: Dataflow support for switch expressions
I am trying to use [switch expressions](http://openjdk.java.net/jeps/325), which is a new [preview feature](http://openjdk.java.net/jeps/12) in Java 12.  My application depends on [NullAway](https://github.com/uber/NullAway), which is a compiler plugin that depends on the Checker Framework's dataflow library.  When I try to compile code that includes a switch expression, NullAway propagates an exception thrown by the dataflow library, causing compilation to fail.

The problematic section of the source code being compiled looks like this:
```java
public static void main(String[] args) {
  String s = switch (args.length) {
    case 0 -> null;
    case 1 -> args[0];
    default -> { throw new IllegalArgumentException("invalid args"); }
  };

  System.out.println(s);
}
```

The root cause of the error that is thrown is this, from the stack trace that NullAway prints:
```
[ERROR]   Caused by: java.lang.AssertionError: case visitor is implemented in SwitchBuilder
[ERROR]         at shadow.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitCase(CFGBuilder.java:3381)
[ERROR]         at shadow.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitCase(CFGBuilder.java:1416)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCase.accept(JCTree.java:1293)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:106)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:114)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.visitSwitchExpression(TreeScanner.java:354)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCSwitchExpression.accept(JCTree.java:1325)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at shadow.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.translateAssignment(CFGBuilder.java:2687)
[ERROR]         at shadow.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitVariable(CFGBuilder.java:4766)
[ERROR]         at shadow.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitVariable(CFGBuilder.java:1416)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:980)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at shadow.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBlock(CFGBuilder.java:3258)
[ERROR]         at shadow.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBlock(CFGBuilder.java:1416)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1038)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
[ERROR]         at shadow.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.process(CFGBuilder.java:1571)
[ERROR]         at shadow.checkerframework.dataflow.cfg.CFGBuilder.build(CFGBuilder.java:255)
[ERROR]         at com.uber.nullaway.dataflow.DataFlow$2.load(DataFlow.java:124)
[ERROR]         at com.uber.nullaway.dataflow.DataFlow$2.load(DataFlow.java:94)
[ERROR]         at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3528)
[ERROR]         at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2277)
[ERROR]         at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2154)
[ERROR]         at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2044)
[ERROR]         ... 110 more
```

That's pointing to this part of the dataflow library: https://github.com/typetools/checker-framework/blob/0e5541ccde8d38ba261d6cc7f670773083e81868/dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java#L3382-L3385

I originally filed [this issue in NullAway's issue tracker](https://github.com/uber/NullAway/issues/289), but the maintainer suggested that it was an upstream issue with the dataflow library so I should also file it here.  I wanted to produce a test case that used dataflow directly without NullAway to share with you, but I couldn't figure out how.

### LINKED ISSUE #4979: Make nodes for switch expressions appear on their own in the containing CFG block
I'm working on getting NullAway to not crash in the presence of switch expressions, now that Checker dataflow does not crash for them.  One small issue is that the marker node for a switch expression may not appear as a top-level node in the containing CFG block, so it never gets directly visited by the dataflow analysis.  If I have a statement `Object o = null;`, I get four nodes in the corresponding CFG basic block: a `VariableDeclarationNode`, a `NullLiteralNode`, an `AssignmentNode`, and a `LocalVariableNode`.  But, if I write `Object o = switch (i) { ... };`, I just get three nodes in the corresponding block:  a `VariableDeclarationNode`, an `AssignmentNode`, and a `LocalVariableNode`; the marker node for the switch expression is missing.

I imagine the representation of `switch` expressions in CFGs is very temporary, and as it gets handled it will get its own node in the containing block for cases like these.  And I can add a special case to work around this for now.  Just wanted to note the issue.

### PR DESCRIPTION
**Pull request by smillst**
They are analyzed the same as switch statements, except that dataflow adds adds an assignment between each result expression and a dataflow-created variable at each yield statement or expression that is a result expression.  The value of that dataflow-created variable is the value of the switch expression.

~~This includes https://github.com/typetools/checker-framework/pull/4978, merge that one first.~~

Delete SKIP-REQUIRE-JAVADOC after merging!

Fixes #2373 and fixes #4979.

### CODE REVIEW

File: SKIP-REQUIRE-JAVADOC [hunk 1]
```diff
@@ -0,0 +1 @@
+Delete after merge.
```
Comment (mernst) [inline SKIP-REQUIRE-JAVADOC:1]: Why is this file necessary?   I don't see wide-ranging uninteresting edits in this pull request.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/SwitchExpressionNode.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+import javax.lang.model.type.TypeMirror;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/** A node for a switch expression. */
+public class SwitchExpressionNode extends Node {
+
+  /** The switch expression tree corresponding to this node. */
```
Comment (mernst) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/SwitchExpressionNode.java:13]: Does this have a specific run-time type?  If so, document that.  Probably the constructor should throw an exception if an object of the wrong type is passed in.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 3]
```diff
@@ -2128,28 +2137,63 @@ public Node visitSwitch(SwitchTree tree, Void p) {
    */
   private class SwitchBuilder {
     /** The switch tree. */
```
Comment (mernst) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:2139]: Javadoc above says this nested class is only for switch statements; I think that should be updated, and this Javadoc should say "It might be a switch statement or a switch expression."

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 4]
```diff
@@ -2128,28 +2137,63 @@ public Node visitSwitch(SwitchTree tree, Void p) {
...
+      this.caseBodyLabels = new Label[caseTrees.size() + 1];
     }
 
-    /** Build up the CFG for the switchTree. */
-    public void build() {
+    /**
+     * Build up the CFG for the switchTree.
+     *
+     * @return if the switch is a switch expression, then a {@link SwitchExpressionNode}; otherwise,
+     *     null
```
Comment (mernst) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:2192]: Why is that the right thing?  I guess probably because elsewhere builds all the CFG nodes for the case bodies and none is needed explicitly for the switch statement itself.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 5]
```diff
@@ -2160,16 +2204,20 @@ public void build() {
...
 
-      // Build CFG for the cases.
-      extendWithNode(
-          new MarkerNode(
-              switchTree,
-              "start of switch statement #" + TreeUtils.treeUids.get(switchTree),
-              env.getTypeUtils()));
+      buildSwitchExpressionVar();
+
+      if (switchTree.getKind() == Kind.SWITCH) {
```
Comment (mernst) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:2209]: After this line, I would add a comment "// It's a switch statement, not a switch expression.".
Likewise elsewhere.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 6]
```diff
@@ -2180,17 +2228,35 @@ public void build() {
...
+                env.getTypeUtils()));
+      }
 
-      extendWithNode(
-          new MarkerNode(
-              switchTree,
-              "end of switch statement #" + TreeUtils.treeUids.get(switchTree),
-              env.getTypeUtils()));
+      switchBuilder = oldSwitchBuilder;
+      if (switchExprVarTree != null) {
```
Comment (mernst) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:2245]: This seems to be the same test as `switchTree.getKind() == Kind.SWITCH` elsewhere.  Would it make the code easier to read & understand if all the tests were consistent?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 7]
```diff
@@ -2256,11 +2341,51 @@ private void buildCase(CaseTree tree, int index) {
           scan(stmt, null);
         }
       } else {
-        scan(TreeUtils.caseTreeGetBody(tree), null);
+        Tree bodyTree = TreeUtils.caseTreeGetBody(tree);
+        if (bodyTree instanceof ExpressionTree) {
```
Comment (mernst) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:2345]: This might be a third different test for the "switch statement vs switch expression" concept.  I think consistency would make the code easier to read.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 8]
```diff
@@ -2256,11 +2341,51 @@ private void buildCase(CaseTree tree, int index) {
...
+        } else {
+          scan(bodyTree, null);
+        }
       }
       extendWithExtendedNode(new UnconditionalJump(nextBodyL));
       addLabelForNextNode(nextCaseL);
     }
+
+    /**
+     * Does the following for each result expression of a switch expression:
```
Comment (mernst) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:2356]: This comment seems wrong.  The client has the logic that this sentence described.  This method only handles one result expression at a time.

**Review by mernst — CHANGES_REQUESTED** [review]

File: SKIP-REQUIRE-JAVADOC (hunk 1 above)
Comment (smillst) [reply SKIP-REQUIRE-JAVADOC]: The problem is that NodeVisitor has 80 undocumented methods.  I would have to document them all to get the Javadoc test to pass. See https://github.com/typetools/checker-framework/runs/4509004521.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java (hunk 4 above)
Comment (smillst) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:2195]: That's right. A switch statement does not have a value so it doesn't need to have a corresponding node.
